### PR TITLE
Better OpenMP

### PR DESCRIPTION
**Pull request by FabioLuporini**
Two main things:

- use `schedule(dynamic,1)` for FD loops (delivers better performance than `schedule(static,1)` on all tested xeons)
- use `schedule(dynamic,chunk_size)` for sparse loops, where `chunk_size` is `number_of_iterations / (nthreads * alpha)` where `alpha` is currently set to 3 (found empirically)

Would appreciate some performance testing

### CODE REVIEW

File: devito/dle/parallelizer.py [hunk 1]
```diff
@@ -90,13 +106,17 @@ class Ompizer(object):
...
-        'for-static': lambda i: c.Pragma('omp for collapse(%d) schedule(static)' % i),
-        'for-static-1': lambda i: c.Pragma('omp for collapse(%d) schedule(static,1)' % i),
-        'for-dynamic-1': lambda i: c.Pragma('omp for collapse(%d) schedule(dynamic,1)'
-                                            % i),
-        'par-for': lambda i, j: c.Pragma('omp parallel for collapse(%d) '
-                                         'schedule(static,1) num_threads(%d)' % (i, j)),
+        'for': lambda i, cs:
+            c.Pragma('omp for collapse(%d) schedule(dynamic,%s)' % (i, cs)),
+        'par-for': lambda i, cs, nt:
+            c.Pragma('omp parallel for collapse(%d) schedule(dynamic,%s) num_threads(%d)'
```
Comment (mloubout) [inline devito/dle/parallelizer.py]: is `dynamic` really always better?

Comment (FabioLuporini) [reply devito/dle/parallelizer.py]: I couldn't find a single use case in which it wasn't .

In fact, I was fairly shocked: both tti and isotropic acoustic on my workstation ran about 10% faster with dynamic

on the KNL is the same story

Comment (mloubout) [reply devito/dle/parallelizer.py]: Wow really, for any threads/space_order/cores...? That's surprising.

Comment (FabioLuporini) [reply devito/dle/parallelizer.py]: well I tried a few, and different grid shapes too...

space orders -- I tried 4 and 16 I think

So not a complete study admittedly, but consistent speedups yeah. Also, in the PR description I asked
> Would appreciate some performance testing

😎

File: devito/ir/iet/nodes.py [hunk 2]
```diff
@@ -129,24 +129,32 @@ def _signature_items(self):
...
 
+    """
+    A mixin for Nodes that represent C expression statements, which are expressions
+    followed by a semicolon. For example, the lines:
 
-class Simple(object):
+        * i = 0;
+        * j = a[i] + 8;
+        * int a = 3;
+        * foo(b)
```
Comment (georgebisbas) [inline devito/ir/iet/nodes.py]: `+;` ?

File: tests/test_operator.py [hunk 3]
```diff
@@ -1513,7 +1513,9 @@ def test_scheduling_sparse_functions(self):
         eqn3 = Eq(u2.forward, u2 + 2*u2.backward - u1.dt2)
         eqn4 = sf2.interpolate(u2)
 
-        op = Operator([eqn1] + eqn2 + [eqn3] + eqn4)
+        # Note: DLE disabled only because with OpenMP otherwise there might be more
```
Comment (georgebisbas) [inline tests/test_operator.py]: A bit of rephrasing?

File: devito/dle/parallelizer.py [hunk 4]
```diff
@@ -22,15 +22,49 @@ def nhyperthreads():
...
 
     @classmethod
     def default_value(cls):
         return int(os.environ.get('OMP_NUM_THREADS', ncores()))
 
     def __new__(cls, **kwargs):
-        return super(NThreads, cls).__new__(cls, name=kwargs['name'], dtype=np.int32,
-                                            value=NThreads.default_value())
+        name = kwargs.get('name', NThreads.name)
+        value = NThreads.default_value()
```
Comment (mloubout) [inline devito/dle/parallelizer.py]: can't pass it in as a kwarg?

File: devito/dle/parallelizer.py [hunk 5]
```diff
@@ -22,15 +22,49 @@ def nhyperthreads():
...
+
+    name = 'nthreads1'
+
+    @classmethod
+    def default_value(cls):
+        return nhyperthreads()
+
+    def __new__(cls, **kwargs):
+        name = kwargs.get('name', NThreadsNested.name)
+        value = NThreadsNested.default_value()
```
Comment (mloubout) [inline devito/dle/parallelizer.py]: same

File: devito/dle/parallelizer.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/dle/parallelizer.py]: no I'm using default_value() for testing too

**Comment by FabioLuporini** [comment]
@georgebisbas :
> Did you try with autotuning off?

yes. 

@mloubout, @georgebisbas 
> branch is green. You've got a few days for your experiments (thanks!). When happy, please poke me and I'll land this

**Comment by mloubout** [comment]
Does seem to improve performance on  my machine (tested mostly 3D tti, rest is cheap :D)

**Comment by FabioLuporini** [comment]
Merged.

**Pull request merged**